File: mobile-client/android/Data/System/Billing/BillingReceiptGateway.kt
package com.smartresponsor.mobile.client.data.system.billing

import com.smartresponsor.mobile.client.contract.system.billing.receiptuploadpayload
import com.smartresponsor.mobile.client.contract.system.billing.receiptverificationresult
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class BillingReceiptGateway(private val baseUrl: String = "https://httpbin.org") {
    private val client = OkHttpClient()

    fun upload(payload: ReceiptUploadPayload): Boolean {
        val body = """{"token":"${payload.token}","product":"${payload.product}"}"""
            .toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url("$baseUrl/anything/mobile/receipt")
            .post(body)
            .build()

        client.newCall(request).execute().use { return it.isSuccessful }
    }

    fun verify(token: String): ReceiptVerificationResult {
        val request = Request.Builder()
            .url("$baseUrl/anything/mobile/verify")
            .post(ByteArray(0).toRequestBody(null))
            .build()

        client.newCall(request).execute().use {
            return ReceiptVerificationResult(accepted = it.isSuccessful)
        }
    }
}

File: client/ios/Data/System/Billing/BillingReceiptGateway.swift
import Foundation

public struct BillingReceiptGateway {
    private let baseUrl: String

    public init(baseUrl: String = "https://httpbin.org") {
        self.baseUrl = baseUrl
    }

    public func upload(payload: ReceiptUploadPayload) async throws -> Bool {
        let url = URL(string: baseUrl + "/anything/mobile/receipt")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = try JSONSerialization.data(
            withJSONObject: ["token": payload.token, "product": payload.product]
        )
        let (_, response) = try await URLSession.shared.data(for: request)
        return (response as? HTTPURLResponse)?.statusCode == 200
    }

    public func verify(token: String) async throws -> ReceiptVerificationResult {
        let url = URL(string: baseUrl + "/anything/mobile/verify")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        let (_, response) = try await URLSession.shared.data(for: request)
        return ReceiptVerificationResult(
            accepted: (response as? HTTPURLResponse)?.statusCode == 200
        )
    }
}
